metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala

```diff
@@ -61,17 +63,28 @@ final class DefinitionProvider(
       token: CancelToken
   ): Future[DefinitionResult] = {
     val fromSemanticdb =
-      semanticdbs.textDocument(path).documentIncludingStale match {
-        case Some(doc) =>
-          definitionFromSnapshot(path, params, doc)
-        case _ =>
-          warnings.noSemanticdb(path)
-          DefinitionResult.empty
+      semanticdbs.textDocument(path).documentIncludingStale
+    val fromSnapshot = fromSemanticdb match {
+      case Some(doc) =>
+        definitionFromSnapshot(path, params, doc)
+      case _ =>
+        DefinitionResult.empty
+    }
+    val fromCompiler =
+      if (fromSnapshot.locations.isEmpty()) {
+        if (remote.isEnabledForPath(path)) {
+          Future(remote.definition(params).getOrElse(fromSnapshot))
```

Comment:
I know `requests` is a synchronous http library (have you considered other async options, by the way?) but wouldn't it make sense for the API for `RemoteLanguageServer` to return futures?

Also, I'm not sure it has a huge impact, but it would probably make sense to mark it as `blocking` as in

```suggestion
          Future(blocking(remote.definition(params).getOrElse(fromSnapshot)))
```

Reply:
I considered using an async client like sttp but chose requests because it has a smaller footprint (total ~100kb). I refactored the `RemoteLanguageServer` API to return futures so that we can replace requests with an async client in the future. I didn't refactor `ReferencesProvider` to become async since it requires changing a large chunk of code and I'm not convinced it's worth the effort right now.

Reply:
Thanks for the explanation. I agree it's probably not worth it to refactor `ReferencesProvider` right now